### PR TITLE
Partial matcher for protos

### PR DESCRIPTION
**Pull request by isovector**
This PR adds a function `partial` which performs a partial match between two protos.

### CODE REVIEW

File: proto-lens-tests/tests/partial_test.hs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    , testAlwaysMatchSame
+    , testMatchRequired
+    , testMatchOptional
+    , testMatchRepeated
+    ]
+
+makeTestCase :: Message a => String -> a -> a -> Test
+makeTestCase testName against match = testCase testName $ do
+    assertBool "matched"         $ partial against match
+    assertBool "not const" . not $ partial match against
```
Comment (judah) [inline proto-lens-tests/tests/partial_test.hs]: What does "not const" mean?

File: proto-lens-tests/tests/partial_test.hs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+testMatchRequired = makeTestCase "required"
+    (def @FooReq & bar . deep . name .~ "deeper")
+    (def @FooReq & ok .~ 5
+                 & bar . name .~ "hello"
+                 & bar . deep . name .~ "deeper")
+
+testMatchOptional = makeTestCase "optional"
+    (def @Foo & ok .~ 5)
+    (def @Foo & ok .~ 5
+                         & bar . name .~ "hello"
```
Comment (judah) [inline proto-lens-tests/tests/partial_test.hs]: Nit: align `&`s with the previous line.

File: proto-lens/src/Data/ProtoLens/Compare.hs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    ) where
+
+import Data.Constraint (Dict (..))
+import qualified Data.Map as Map
+import Lens.Family2 (Lens', over, set, view)
+import Lens.Family2.Unchecked (lens)
+import Data.ProtoLens.Message
+
+-- | Perform a partial match on a protobuffer. This corresponds to equality for
+-- primitive fields, (recursive) partial equality for required 'Message'
```
Comment (judah) [inline proto-lens/src/Data/ProtoLens/Compare.hs]: We should probably be more explicit in the comment: this is equality for proto3 scalar optional fields and proto2 required fields.

File: proto-lens/src/Data/ProtoLens/Compare.hs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+-- fields, partial equality if both are set for optional fields, and partial
+-- equality of subsets for repeated and map fields.
```
Comment (judah) [inline proto-lens/src/Data/ProtoLens/Compare.hs]: I have mixed feelings about subset equality for repeated fields and whether that satisfies the principle of least surprise.  I can recall a couple tests I've written where I did want to specify the explicit size/order of the list, I just didn't care about every field of those values.

File: proto-lens/src/Data/ProtoLens/Compare.hs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+        RepeatedField _ l ->
+          and . fmap (\a -> any (eq ftd a) $ view l match) $ view l against
+
+        MapField k v l    ->
+          let entriesMatch   = toRepeatedEntry k v $ view l match
+              entriesAgainst = toRepeatedEntry k v $ view l against
+           in and $ fmap (\a -> any (eq ftd a) entriesMatch) entriesAgainst
+  where
+    toRepeatedEntry :: (Ord key, Message entry)
```
Comment (judah) [inline proto-lens/src/Data/ProtoLens/Compare.hs]: Break these helper functions out into separate, top-level (unexported) functions.

Ok if it makes sense to group them together; for example, make `getEq` part of the where clause of `eq`.

File: proto-lens-tests/tests/partial_test.hs (hunk 1 above)
Comment (isovector) [reply proto-lens-tests/tests/partial_test.hs]: it means `partial` isn't implemented as `\_ _ -> True`  :)

File: proto-lens/src/Data/ProtoLens/Compare.hs (hunk 4 above)
Comment (isovector) [reply proto-lens/src/Data/ProtoLens/Compare.hs]: yeah, iirc you somehow compose matchers at google to decide which of these behaviors you want

File: proto-lens/src/Data/ProtoLens/Compare.hs [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+import qualified Data.Map as Map
+import Lens.Family2 (Lens', over, set, view)
+import Lens.Family2.Unchecked (lens)
+import Data.ProtoLens.Message
+
+-- | Perform a partial match on a protobuffer. This corresponds to equality for
+-- primitive fields, (recursive) partial equality for required 'Message'
+-- fields, partial equality if both are set for optional fields, and partial
+-- equality of subsets for repeated and map fields.
+partial
```
Comment (judah) [inline proto-lens/src/Data/ProtoLens/Compare.hs]: One other request: can we call this something more specific like `partialEq`?  The word `partial` has an overloaded meaning (e.g. a function like `head`)...

**Comment by judah** [comment]
Looks like this failed in CI for a couple reasons:
1. Warnings; please re-run with `stack test --pedantic`
2. `TypeApplications` isn't supported in 7.10.  It might be enough to just wait for #136.

**Comment by judah** [comment]
Closing since this PR has become pretty stale.

**Pull request closed without merging**